bundler/lib/bundler/plugin.rb

```diff
@@ -100,23 +102,28 @@ def list
     #
     # @param [Pathname] gemfile path
     # @param [Proc] block that can be evaluated for (inline) Gemfile
-    def gemfile_install(gemfile = nil, &inline)
-      Bundler.settings.temporary(frozen: false, deployment: false) do
+    def gemfile_install(gemfile = nil, lockfile = nil, unlock = {}, &inline)
+      # skip the update if unlocking specific gems, but none of them are our plugins
+      if unlock.is_a?(Hash) && unlock[:gems] && !unlock[:gems].empty? &&
+         (unlock[:gems] & index.installed_plugins).empty?
+        unlock = {}
+      end
+
```

Comment:
Except that even if unlocking other gems, this pass through the Gemfile doesn't even record the other dependencies, _only_ the plugins. So that's simply not possible. On the second pass through the gemfile, it may be possible for the plugin's version to change due to other gems' dependencies. So I think there might be a bug there, but the fix would involve re-installing plugins _again_ after the update command completes, rebuilding it based on the full lockfile. Removing this optimization would do nothing to fix that issue, and it's pretty big performance thing if you commonly use `bundle update <specific gem>`, since it means it will have to contact RubyGems _twice_ for every one of those commands.

Reply:
Oh, that makes sense. I actually thought the first pass was the "standard one", and the second pass was plugins, but it seems to be actually the other way around? Would inverting the order fix this potential issue?

Reply:
Plugins is the first pass, and can't be made the second--regular--pass, since the second pass needs to have plugins loaded that were installed in the first pass.

Reply:
I see. So this is because of those "custom `:type` sources" defined as plugin gems, right? In my opinion, those plugins are special and should be installed manually and globally before bundler is run. Then everything else could be unified and we could instantiate and install a single definition. That'd be probably going too far and breaking, but maybe we could leave the "first phase" only for installing these extra sources (like `definition.add_custom_sources` or something) and don't add them to the lockfile, then unify everything else. Basically, if these extra special gems are necessary to properly parse the `Gemfile`, I'm not sure if it makes sense to include them in the lockfile.

Overall, I think we'll eventually want a better unified solution, but happy to punt on this here and leave it for another time, since I think this approach is still much better than what we have now.

Reply:
Fair enough, I agree 👍.